File: composeApp/src/androidMain/kotlin/com/kissangram/viewmodel/HomeViewModel.kt
package com.kissangram.viewmodel

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.kissangram.model.Post
import com.kissangram.model.UserStories
import com.kissangram.repository.AndroidAuthRepository
import com.kissangram.repository.AndroidPreferencesRepository
import com.kissangram.repository.FirestoreFeedRepository
import com.kissangram.repository.FirestoreStoryRepository
import com.kissangram.repository.FirestorePostRepository
import com.kissangram.repository.FirestoreUserRepository
import com.kissangram.usecase.GetHomeFeedUseCase
import com.kissangram.usecase.GetStoryBarUseCase
import com.kissangram.usecase.LikePostUseCase
import com.kissangram.usecase.SavePostUseCase
import android.util.Log
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.Timestamp
import com.kissangram.data.LocationData
import com.kissangram.data.CropsData

class HomeViewModel(
    application: Application
) : AndroidViewModel(application) {

    private val authRepository = AndroidAuthRepository(
        context = application,
        activity = null,
        preferencesRepository = AndroidPreferencesRepository(application)
    )
    private val userRepository = FirestoreUserRepository(authRepository = authRepository)
    private val feedRepository = FirestoreFeedRepository(authRepository = authRepository)
    private val postRepository = FirestorePostRepository(
        authRepository = authRepository,
        userRepository = userRepository
    )
    private val storyRepository = FirestoreStoryRepository()
    
    // Use cases
    private val getHomeFeedUseCase = GetHomeFeedUseCase(feedRepository)
    private val getStoryBarUseCase = GetStoryBarUseCase(storyRepository)
    private val likePostUseCase = LikePostUseCase(postRepository)
    private val savePostUseCase = SavePostUseCase(postRepository)
    
    private val _uiState = MutableStateFlow(HomeUiState())
    val uiState: StateFlow<HomeUiState> = _uiState.asStateFlow()
    
    // Track posts currently being processed to prevent race conditions
    private val postsBeingProcessed = mutableSetOf<String>()
    
    init {
        Log.d(TAG, "init: HomeViewModel created, calling loadContent()")
        loadContent()
    }
    
    fun loadContent() {
        viewModelScope.launch {
            Log.d(TAG, "loadContent: start, setting isLoading=true")
            _uiState.value = _uiState.value.copy(isLoading = true, error = null)
            val userId = authRepository.getCurrentUserId()
            Log.d(TAG, "loadContent: currentUserId=${userId?.take(8)?.plus("...") ?: "null"}")
            try {
                val stories = getStoryBarUseCase()
                Log.d(TAG, "loadContent: stories count=${stories.size}")
                val posts = getHomeFeedUseCase(page = 0)
                Log.d(TAG, "loadContent: posts count=${posts.size}, firstId=${posts.firstOrNull()?.id}")
                _uiState.value = _uiState.value.copy(
                    stories = stories,
                    posts = posts,
                    isLoading = false
                )
                Log.d(TAG, "loadContent: success, uiState.posts.size=${_uiState.value.posts.size}")
            } catch (e: Exception) {
                Log.e(TAG, "loadContent: failed", e)
                _uiState.value = _uiState.value.copy(
                    isLoading = false,
                    error = e.message ?: "Failed to load content"
                )
            }
        }
    }
    
    fun refreshFeed() {
        viewModelScope.launch {
            Log.d(TAG, "refreshFeed: start")
            _uiState.value = _uiState.value.copy(isRefreshing = true)
            try {
                val stories = getStoryBarUseCase()
                val posts = getHomeFeedUseCase(page = 0)
                Log.d(TAG, "refreshFeed: stories=${stories.size}, posts=${posts.size}")
                _uiState.value = _uiState.value.copy(
                    stories = stories,
                    posts = posts,
                    isRefreshing = false
                )
            } catch (e: Exception) {
                Log.e(TAG, "refreshFeed: failed", e)
                _uiState.value = _uiState.value.copy(
                    isRefreshing = false,
                    error = e.message
                )
            }
        }
    }
    
    fun loadMorePosts() {
        if (_uiState.value.isLoadingMore) return
        viewModelScope.launch {
            Log.d(TAG, "loadMorePosts: start, currentPage=${_uiState.value.currentPage}")
            _uiState.value = _uiState.value.copy(isLoadingMore = true)
            try {
                val currentPage = _uiState.value.currentPage
                val newPosts = getHomeFeedUseCase(page = currentPage + 1)
                Log.d(TAG, "loadMorePosts: got ${newPosts.size} new posts")
                if (newPosts.isNotEmpty()) {
                    _uiState.value = _uiState.value.copy(
                        posts = _uiState.value.posts + newPosts,
                        currentPage = currentPage + 1,
                        isLoadingMore = false
                    )
                } else {
                    _uiState.value = _uiState.value.copy(
                        hasMorePosts = false,
                        isLoadingMore = false
                    )
                }
            } catch (e: Exception) {
                Log.e(TAG, "loadMorePosts: failed", e)
                _uiState.value = _uiState.value.copy(isLoadingMore = false)
            }
        }
    }
    
    /**
     * Handles like/unlike action for a post.
     * @return true if the request was accepted and processed, false if it was ignored (e.g., already processing)
     */
    fun onLikePost(postId: String): Boolean {
        // Prevent multiple simultaneous requests for the same post
        if (postsBeingProcessed.contains(postId)) {
            Log.d(TAG, "onLikePost: post $postId is already being processed, ignoring")
            return false
        }
        
        val currentPosts = _uiState.value.posts
        val postIndex = currentPosts.indexOfFirst { it.id == postId }
        
        if (postIndex != -1) {
            val post = currentPosts[postIndex]
            val newLikedState = !post.isLikedByMe
            val newLikesCount = if (newLikedState) post.likesCount + 1 else post.likesCount - 1
            
            // Mark as being processed
            postsBeingProcessed.add(postId)
            
            // ⚡ INSTAGRAM APPROACH: Update UI IMMEDIATELY (synchronous, main thread)
            // This happens before any async work, so user sees instant feedback
            val updatedPost = post.copy(
                isLikedByMe = newLikedState,
                likesCount = newLikesCount
            )
            val updatedPosts = currentPosts.toMutableList()
            updatedPosts[postIndex] = updatedPost
            _uiState.value = _uiState.value.copy(posts = updatedPosts)
            
            // Fire network request in background (non-blocking)
            viewModelScope.launch {
                try {
                    likePostUseCase(postId, post.isLikedByMe)
                } catch (e: Exception) {
                    Log.e(TAG, "onLikePost: failed for postId=$postId", e)
                    // Revert on failure
                    val revertedPosts = currentPosts.toMutableList()
                    _uiState.value = _uiState.value.copy(posts = revertedPosts)
                } finally {
                    // Always remove from processing set
                    postsBeingProcessed.remove(postId)
                }
            }
            return true
        }
        return false
    }
    
    fun onSavePost(postId: String) {
        // Prevent multiple simultaneous requests for the same post
        val saveKey = "save_$postId"
        if (postsBeingProcessed.contains(saveKey)) {
            Log.d(TAG, "onSavePost: post $postId is already being processed, ignoring")
            return
        }
        
        val currentPosts = _uiState.value.posts
        val postIndex = currentPosts.indexOfFirst { it.id == postId }
        
        if (postIndex != -1) {
            val post = currentPosts[postIndex]
            val newSavedState = !post.isSavedByMe
            
            // Mark as being processed
            postsBeingProcessed.add(saveKey)
            
            // ⚡ INSTAGRAM APPROACH: Update UI IMMEDIATELY (synchronous, main thread)
            // This happens before any async work, so user sees instant feedback
            val updatedPost = post.copy(isSavedByMe = newSavedState)
            val updatedPosts = currentPosts.toMutableList()
            updatedPosts[postIndex] = updatedPost
            _uiState.value = _uiState.value.copy(posts = updatedPosts)
            
            // Fire network request in background (non-blocking)
            viewModelScope.launch {
                try {
                    savePostUseCase(postId, post.isSavedByMe)
                } catch (e: Exception) {
                    Log.e(TAG, "onSavePost: failed for postId=$postId", e)
                    // Revert on failure
                    val revertedPosts = currentPosts.toMutableList()
                    _uiState.value = _uiState.value.copy(posts = revertedPosts)
                } finally {
                    // Always remove from processing set
                    postsBeingProcessed.remove(saveKey)
                }
            }
        }
    }
    
    /**
     * Upload India states and districts data to Firestore.
     * This is a one-time operation for seeding reference data.
     * Call this from a dev button, then remove the button after use.
     */
    fun uploadLocationsToFirestore(
        onSuccess: () -> Unit,
        onError: (String) -> Unit
    ) {
        viewModelScope.launch {
            try {
                val firestore = FirebaseFirestore.getInstance(
        com.google.firebase.FirebaseApp.getInstance(),
       "kissangram"
    )
                
                val locationData = hashMapOf(
                    "country" to "India",
                    "statesAndDistricts" to LocationData.indiaStatesAndDistricts,
                    "stateNames" to LocationData.stateNames,
                    "version" to 1,
                    "updatedAt" to Timestamp.now()
                )
                
                firestore.collection("appConfig")
                    .document("locations")
                    .set(locationData)
                    .addOnSuccessListener {
                        onSuccess()
                    }
                    .addOnFailureListener { e ->
                        onError(e.message ?: "Upload failed")
                    }
            } catch (e: Exception) {
                onError(e.message ?: "Failed to upload locations")
            }
        }
    }
    
    /**
     * Upload crops data to Firestore.
     * This is a one-time operation for seeding reference data.
     * Call this from a dev button, then remove the button after use.
     */
    fun uploadCropsToFirestore(
        onSuccess: () -> Unit,
        onError: (String) -> Unit
    ) {
        viewModelScope.launch {
            try {
                val firestore = FirebaseFirestore.getInstance(
                    com.google.firebase.FirebaseApp.getInstance(),
                    "kissangram"
                )
                
                val cropsData = hashMapOf(
                    "categories" to CropsData.categories,
                    "categoryNames" to CropsData.categoryNames,
                    "allCrops" to CropsData.allCrops,
                    "totalCrops" to CropsData.totalCrops,
                    "version" to 1,
                    "updatedAt" to Timestamp.now()
                )
                
                firestore.collection("appConfig")
                    .document("crops")
                    .set(cropsData)
                    .addOnSuccessListener {
                        onSuccess()
                    }
                    .addOnFailureListener { e ->
                        onError(e.message ?: "Upload failed")
                    }
            } catch (e: Exception) {
                onError(e.message ?: "Failed to upload crops")
            }
        }
    }

    companion object {
        private const val TAG = "HomeViewModel"
    }
}

data class HomeUiState(
    val stories: List<UserStories> = emptyList(),
    val posts: List<Post> = emptyList(),
    val isLoading: Boolean = false,
    val isRefreshing: Boolean = false,
    val isLoadingMore: Boolean = false,
    val hasMorePosts: Boolean = true,
    val currentPage: Int = 0,
    val error: String? = null
)

File: iosApp/iosApp/viewmodel/HomeViewModel.swift
import Foundation
import os.log
import Shared

@MainActor
class HomeViewModel: ObservableObject {

    private static let log = Logger(subsystem: "com.kissangram", category: "HomeViewModel")
    private let prefs = IOSPreferencesRepository()
    private let authRepository: AuthRepository
    private let followRepository: FollowRepository
    private let feedRepository: FeedRepository
    private let postRepository: PostRepository
    private let storyRepository: StoryRepository

    private let getHomeFeedUseCase: GetHomeFeedUseCase
    private let followUserUseCase: FollowUserUseCase
    private let getStoryBarUseCase: GetStoryBarUseCase
    private let likePostUseCase: LikePostUseCase
    private let savePostUseCase: SavePostUseCase

    @Published var stories: [UserStories] = []
    @Published var posts: [Post] = []
    @Published var authorIdToIsFollowing: [String: Bool] = [:]
    @Published var currentUserId: String?
    @Published var isLoading = false
    @Published var isRefreshing = false
    @Published var isLoadingMore = false
    @Published var hasMorePosts = true
    @Published var error: String?

    private var currentPage: Int32 = 0
    
    // Track posts currently being processed to prevent race conditions
    private var postsBeingProcessed = Set<String>()

    init() {
        self.authRepository = IOSAuthRepository(preferencesRepository: prefs)
        let userRepository = FirestoreUserRepository(authRepository: authRepository)
        self.followRepository = IOSFollowRepository(authRepository: authRepository, userRepository: userRepository)
        self.feedRepository = FirestoreFeedRepository(authRepository: authRepository)
        self.postRepository = FirestorePostRepository(authRepository: authRepository, userRepository: userRepository)
        self.storyRepository = FirestoreStoryRepository(authRepository: authRepository, followRepository: followRepository)

        self.getHomeFeedUseCase = GetHomeFeedUseCase(feedRepository: feedRepository)
        self.getStoryBarUseCase = GetStoryBarUseCase(storyRepository: storyRepository)
        self.followUserUseCase = FollowUserUseCase(followRepository: followRepository)
        self.likePostUseCase = LikePostUseCase(postRepository: postRepository)
        self.savePostUseCase = SavePostUseCase(postRepository: postRepository)

        Self.log.debug("HomeViewModel init, calling loadContent")
        Task {
            await loadContent()
        }
    }
    
    func loadContent() async {
        isLoading = true
        error = nil
        Self.log.debug("loadContent: start")
        do {
            async let storiesResult = storyRepository.getStoryBar()
            async let postsResult = feedRepository.getHomeFeed(page: 0, pageSize: 20, forceRefresh: false)
            
            let (loadedStories, loadedPosts) = try await (storiesResult, postsResult)
            
            Self.log.debug("loadContent: stories=\(loadedStories.count) posts=\(loadedPosts.count) firstPostId=\(loadedPosts.first?.id ?? "nil")")
            let userId = try? await authRepository.getCurrentUserId()
            var followingMap: [String: Bool] = [:]
            for authorId in Set(loadedPosts.map { $0.authorId }).filter({ $0 != userId }) {
                if let isFollowing = try? await followRepository.isFollowing(userId: authorId) {
                    followingMap[authorId] = isFollowing.boolValue
                }
            }
            self.stories = loadedStories
            self.posts = loadedPosts
            self.authorIdToIsFollowing = followingMap
            self.currentUserId = userId
            self.currentPage = 0
            self.isLoading = false
            Self.log.debug("loadContent: success")
        } catch {
            Self.log.error("loadContent: error=\(error.localizedDescription)")
            self.isLoading = false
            self.error = error.localizedDescription
        }
    }
    
    func refreshFeed() async {
        isRefreshing = true
        Self.log.debug("refreshFeed: start")
        do {
            async let storiesResult = storyRepository.getStoryBar()
            async let postsResult = feedRepository.getHomeFeed(page: 0, pageSize: 20, forceRefresh: true)
            
            let (loadedStories, loadedPosts) = try await (storiesResult, postsResult)
            Self.log.debug("refreshFeed: stories=\(loadedStories.count) posts=\(loadedPosts.count)")
            let userId = try? await authRepository.getCurrentUserId()
            var followingMap: [String: Bool] = [:]
            for authorId in Set(loadedPosts.map { $0.authorId }).filter({ $0 != userId }) {
                if let isFollowing = try? await followRepository.isFollowing(userId: authorId) {
                    followingMap[authorId] = isFollowing.boolValue
                }
            }
            self.stories = loadedStories
            self.posts = loadedPosts
            self.authorIdToIsFollowing = followingMap
            self.currentUserId = userId
            self.currentPage = 0
            self.isRefreshing = false
        } catch {
            Self.log.error("refreshFeed: error=\(error.localizedDescription)")
            self.isRefreshing = false
        }
    }
    
    func loadMorePosts() async {
        guard !isLoadingMore && hasMorePosts else { return }
        isLoadingMore = true
        let page = currentPage + 1
        Self.log.debug("loadMorePosts: currentPage=\(self.currentPage) nextPage=\(page)")
        do {
            let newPosts = try await feedRepository.getHomeFeed(page: page, pageSize: 20, forceRefresh: false)
            Self.log.debug("loadMorePosts: newPosts=\(newPosts.count)")
            if !newPosts.isEmpty {
                let userId = try? await authRepository.getCurrentUserId()
                var newFollowingMap = self.authorIdToIsFollowing
                for authorId in Set(newPosts.map { $0.authorId }).filter({ $0 != userId && !newFollowingMap.keys.contains($0) }) {
                    if let isFollowing = try? await followRepository.isFollowing(userId: authorId) {
                        newFollowingMap[authorId] = isFollowing.boolValue
                    }
                }
                self.posts.append(contentsOf: newPosts)
                self.authorIdToIsFollowing = newFollowingMap
                self.currentPage += 1
            } else {
                self.hasMorePosts = false
            }
            self.isLoadingMore = false
        } catch {
            Self.log.error("loadMorePosts: error=\(error.localizedDescription)")
            self.isLoadingMore = false
        }
    }
    
    /**
     * Handles like/unlike action for a post.
     * @return true if the request was accepted and processed, false if it was ignored (e.g., already processing)
     */
    func onLikePost(_ postId: String) -> Bool {
        // Prevent multiple simultaneous requests for the same post
        guard !postsBeingProcessed.contains(postId) else {
            Self.log.debug("onLikePost: post \(postId) is already being processed, ignoring")
            return false
        }
        
        guard let index = posts.firstIndex(where: { $0.id == postId }) else { return false }
        
        let post = posts[index]
        let newLikedState = !post.isLikedByMe
        let newLikesCount = newLikedState ? post.likesCount + 1 : post.likesCount - 1
        
        // Mark as being processed
        postsBeingProcessed.insert(postId)
        
        // ⚡ INSTAGRAM APPROACH: Update UI IMMEDIATELY (synchronous, main thread)
        // Since we're @MainActor, this happens synchronously before any async work
        // User sees instant feedback with zero perceived lag
        posts[index] = Post(
            id: post.id,
            authorId: post.authorId,
            authorName: post.authorName,
            authorUsername: post.authorUsername,
            authorProfileImageUrl: post.authorProfileImageUrl,
            authorRole: post.authorRole,
            authorVerificationStatus: post.authorVerificationStatus,
            type: post.type,
            text: post.text,
            media: post.media,
            voiceCaption: post.voiceCaption,
            crops: post.crops,
            hashtags: post.hashtags,
            location: post.location,
            question: post.question,
            likesCount: newLikesCount,
            commentsCount: post.commentsCount,
            savesCount: post.savesCount,
            isLikedByMe: newLikedState,
            isSavedByMe: post.isSavedByMe,
            createdAt: post.createdAt,
            updatedAt: post.updatedAt
        )
        
        Task {
            defer {
                // Always remove from processing set
                postsBeingProcessed.remove(postId)
            }
            
            do {
                try await likePostUseCase.invoke(postId: postId, isCurrentlyLiked: post.isLikedByMe)
            } catch {
                Self.log.error("onLikePost: failed for postId=\(postId), error=\(error.localizedDescription)")
                // Revert on failure
                self.posts[index] = post
            }
        }
        
        return true
    }
    
    func onSavePost(_ postId: String) {
        // Prevent multiple simultaneous requests for the same post
        let saveKey = "save_\(postId)"
        guard !postsBeingProcessed.contains(saveKey) else {
            Self.log.debug("onSavePost: post \(postId) is already being processed, ignoring")
            return
        }
        
        guard let index = posts.firstIndex(where: { $0.id == postId }) else { return }
        
        let post = posts[index]
        let newSavedState = !post.isSavedByMe
        
        // Mark as being processed
        postsBeingProcessed.insert(saveKey)
        
        // ⚡ INSTAGRAM APPROACH: Update UI IMMEDIATELY (synchronous, main thread)
        // Since we're @MainActor, this happens synchronously before any async work
        // User sees instant feedback with zero perceived lag
        posts[index] = Post(
            id: post.id,
            authorId: post.authorId,
            authorName: post.authorName,
            authorUsername: post.authorUsername,
            authorProfileImageUrl: post.authorProfileImageUrl,
            authorRole: post.authorRole,
            authorVerificationStatus: post.authorVerificationStatus,
            type: post.type,
            text: post.text,
            media: post.media,
            voiceCaption: post.voiceCaption,
            crops: post.crops,
            hashtags: post.hashtags,
            location: post.location,
            question: post.question,
            likesCount: post.likesCount,
            commentsCount: post.commentsCount,
            savesCount: post.savesCount,
            isLikedByMe: post.isLikedByMe,
            isSavedByMe: newSavedState,
            createdAt: post.createdAt,
            updatedAt: post.updatedAt
        )
        
        Task {
            defer {
                // Always remove from processing set
                postsBeingProcessed.remove(saveKey)
            }
            
            do {
                try await savePostUseCase.invoke(postId: postId, isCurrentlySaved: post.isSavedByMe)
            } catch {
                Self.log.error("onSavePost: failed for postId=\(postId), error=\(error.localizedDescription)")
                // Revert on failure
                self.posts[index] = post
            }
        }
    }

    func onFollow(authorId: String) {
        Task {
            do {
                try await followUserUseCase.invoke(userId: authorId, isCurrentlyFollowing: false)
                await MainActor.run {
                    authorIdToIsFollowing[authorId] = true
                }
            } catch {
                Self.log.error("onFollow: failed for authorId=\(authorId)")
            }
        }
    }

    func unfollowAndRemovePosts(authorId: String) {
        Task {
            do {
                try await followUserUseCase.invoke(userId: authorId, isCurrentlyFollowing: true)
                await MainActor.run {
                    posts = posts.filter { $0.authorId != authorId }
                    authorIdToIsFollowing[authorId] = false
                }
            } catch {
                Self.log.error("unfollowAndRemovePosts: failed for authorId=\(authorId)")
            }
        }
    }
}
